src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -96,9 +96,23 @@ protected void ShutdownConnectedNodes(List<NodeContext> contextsToShutDown, bool
             // Send the build completion message to the nodes, causing them to shutdown or reset.
             _processesToIgnore.Clear();
 
+            // We wait for child nodes to exit to avoid them changing the terminal
+            // after this process terminates.
+            bool waitForExit =  !enableReuse &&
+                                !Console.IsInputRedirected &&
+                                Traits.Instance.EscapeHatches.EnsureStdOutForChildNodesIsPrimaryStdout;
+
             foreach (NodeContext nodeContext in contextsToShutDown)
             {
-                nodeContext?.SendData(new NodeBuildComplete(enableReuse));
+                if (nodeContext is null)
+                {
+                    continue;
+                }
+                nodeContext.SendData(new NodeBuildComplete(enableReuse));
+                if (waitForExit)
```

Comment:
I think you have to add a `bool waitForExit` parameter to the method. I looked at the callers of this method and some expect this to be async:

![image](https://user-images.githubusercontent.com/2255729/105433550-ab68c600-5c0e-11eb-8e3e-83695f5505b7.png)

Reply:
ShutdownConnectedNodesAsync doesn't seem to have the async keyword. Am I missing something?

Reply:
I think it's from before C# had async / await, so its meaning is more like `ShutdownConnectedNodesWithoutWaiting`.

Reply:
I think in case of `ShutdownConnectedNodesAsync(bool abort)` with `abort:true` we may want to `Kill` the children, and still `WaitForExit`. Does that make sense?

Reply:
I've added some of the `Kill` logic into the `WaitForExitAsync` method, and a timeout.

I haven't added an extra parameter to the method to indicate the timeout should be skipped and the node should be killed immediately. Do you think it is still needed?

Reply:
Regarding `ShutdownConnectedNodesAsync`, the method name becomes misleading if nodes are now waited. At the least remove `Async` from the method name. Though I do worry that it might trigger RPS tests around VS shutdown time. @Forgind could you please trigger an RPS run for the PR?